chain/network/src/peer.rs

```diff
@@ -61,10 +111,19 @@ impl Peer {
             handshake_timeout,
             peer_manager_addr,
             client_addr,
+            tracker: Default::default(),
         }
     }
 
     fn send_message(&mut self, msg: PeerMessage) {
+        // Skip sending block and headers if we received it or header from this peer.
+        // Record block requests in tracker.
+        match msg {
+            PeerMessage::Block(b) if self.tracking.has_received(b.hash()) => return,
+            PeerMessage::BlockHeaderAnnounrce(h) if self.tracking.has_received(h.hash()) => return,
+            PeerMessage::BlockRequest(h) => self.tracker.push_request(h),
```

Comment:
When it sends blocks, why doesn't it push the hash to tracker?

Reply:
Because it's not a request or recevied?